File: Tableau-Supported/Python/read_and_print_data_from_existing_hyper_file.py
# -----------------------------------------------------------------------------
#
# This file is the copyrighted property of Tableau Software and is protected
# by registered patents and other applicable U.S. and international laws and
# regulations.
#
# You may adapt this file and modify it to fit into your context and use it
# as a template to start your own projects.
#
# -----------------------------------------------------------------------------
import shutil

from pathlib import Path

from tableauhyperapi import HyperProcess, Telemetry, \
    Connection, CreateMode, \
    NOT_NULLABLE, NULLABLE, SqlType, TableDefinition, \
    Inserter, \
    escape_name, escape_string_literal, \
    TableName, \
    HyperException


def run_read_data_from_existing_hyper_file():
    """
    An example of how to read and print data from an existing Hyper file.
    """
    print("EXAMPLE - Read data from an existing Hyper file")

    # Path to a Hyper file containing all data inserted into Customer, Product, Orders and LineItems table.
    # See "insert_data_into_multiple_tables.py" for an example that works with the complete schema.
    path_to_source_database = Path(__file__).parent / "data" / "superstore_sample_denormalized.hyper"

    # Make a copy of the superstore denormalized sample Hyper file
    path_to_database = Path(shutil.copy(src=path_to_source_database,
                                        dst="superstore_sample_denormalized_read.hyper")).resolve()

    # Starts the Hyper Process with telemetry enabled to send data to Tableau.
    # To opt out, simply set telemetry=Telemetry.DO_NOT_SEND_USAGE_DATA_TO_TABLEAU.
    with HyperProcess(telemetry=Telemetry.SEND_USAGE_DATA_TO_TABLEAU) as hyper:

        # Connect to existing Hyper file "superstore_sample_denormalized_read.hyper".
        with Connection(endpoint=hyper.endpoint,
                        database=path_to_database) as connection:
            # The table names in the "Extract" schema (the default schema).
            table_names = connection.catalog.get_table_names(schema="Extract")

            for table in table_names:
                table_definition = connection.catalog.get_table_definition(name=table)
                print(f"Table {table.name} has qualified name: {table}")
                for column in table_definition.columns:
                    print(f"Column {column.name} has type={column.type} and nullability={column.nullability}")
                print("")

            # Print all rows from the "Extract"."Extract" table.
            table_name = TableName("Extract", "Extract")
            print(f"These are all rows in the table {table_name}:")
            # `execute_list_query` executes a SQL query and returns the result as list of rows of data,
            # each represented by a list of objects.
            rows_in_table = connection.execute_list_query(query=f"SELECT * FROM {table_name}")
            print(rows_in_table)

        print("The connection to the Hyper file has been closed.")
    print("The Hyper process has been shut down.")


if __name__ == '__main__':
    try:
        run_read_data_from_existing_hyper_file()
    except HyperException as ex:
        print(ex)
        exit(1)

File: Tableau-Supported/CPP/read_and_print_data_from_existing_hyper_file.cpp
// -----------------------------------------------------------------------------
//
// This file is the copyrighted property of Tableau Software and is protected
// by registered patents and other applicable U.S. and international laws and
// regulations.
//
// You may adapt this file and modify it to fit into your context and use it
// as a template to start your own projects.
//
// -----------------------------------------------------------------------------

/**
 * \example read_and_print_data_from_existing_hyper_file.cpp
 *
 * An example of how to read and print data from an existing Hyper file.
 */

#include <fstream>
#include <hyperapi/hyperapi.hpp>
#include <iostream>
#include <string>
#include <unordered_set>

/**
 * Helper function to copy a file
 */
static void copy(const std::string& sourcePath, const std::string& destinationPath) {
   std::ifstream source(sourcePath, std::ios::binary);
   std::ofstream destination(destinationPath, std::ios::binary);
   destination << source.rdbuf();
   source.close();
   destination.close();
}

static void runReadAndPrintDataFromExistingHyperFile() {
   std::cout << "EXAMPLE - Read data from an existing Hyper file" << std::endl;

   // Path to a Hyper file containing all data inserted into "Extract"."Extract" table
   // See "insert_data_into_single_table.cpp" for an example that works with the complete schema.
   const std::string pathToSourceDatabase = "data/superstore_sample_denormalized.hyper";

   // Make a copy of the superstore example Hyper file.
   const std::string pathToDatabase = "data/superstore_sample_denormalized_read.hyper";
   copy(pathToSourceDatabase, pathToDatabase);

   // Starts the Hyper Process with telemetry enabled to send data to Tableau.
   // To opt out, simply set telemetry=hyperapi::Telemetry::DoNotSendUsageDataToTableau.
   {
      hyperapi::HyperProcess hyper(hyperapi::Telemetry::SendUsageDataToTableau);

      // Connect to existing Hyper file "superstore_sample_denormalized_read.hyper".
      {
         hyperapi::Connection connection(hyper.getEndpoint(), pathToDatabase);
         const hyperapi::Catalog& catalog = connection.getCatalog();

         // The table names in the "Extract" schema.
         std::unordered_set<hyperapi::TableName> tableNames = catalog.getTableNames("Extract");
         for (auto& tableName : tableNames) {
            hyperapi::TableDefinition tableDefinition = catalog.getTableDefinition(tableName);
            std::cout << "Table " << tableName << " has qualified name: " << tableDefinition.getTableName() << std::endl;
            for (auto& column : tableDefinition.getColumns()) {
               std::cout << "\t Column " << column.getName() << " has type " << column.getType() << " and nullability " << column.getNullability()
                         << std::endl;
            }
            std::cout << std::endl;
         }

         // Print all rows from the "Extract"."Extract" table.
         hyperapi::TableName extractTable("Extract", "Extract");
         std::cout << "These are all rows in the table " << extractTable.toString() << ":" << std::endl;

         hyperapi::Result rowsInTable = connection.executeQuery("SELECT * FROM " + extractTable.toString());
         for (const hyperapi::Row& row : rowsInTable) {
            for (const hyperapi::Value& value : row) {
               std::cout << value << '\t';
            }
            std::cout << '\n';
         }
      }
      std::cout << "The connection to the Hyper file has been closed." << std::endl;
   }
   std::cout << "The Hyper Process has been shut down." << std::endl;
}

int main() {
   try {
      runReadAndPrintDataFromExistingHyperFile();
   } catch (const hyperapi::HyperException& e) {
      std::cout << e.toString() << std::endl;
      return 1;
   }
   return 0;
}
